src/backend/columnar/columnar_customscan.c

```diff
@@ -542,9 +895,20 @@ CreateColumnarScanPath(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
  * need to be read.
  */
 static Cost
-ColumnarScanCost(RelOptInfo *rel, Oid relationId, int numberOfColumnsRead)
+ColumnarScanCost(RelOptInfo *rel, Oid relationId, int numberOfColumnsRead, int nClauses)
 {
-	return ColumnarTableStripeCount(relationId) *
+	double stripesToRead = ColumnarTableStripeCount(relationId);
+
+	/*
+	 * XXX: we don't currently have a good way of estimating the selectivity
+	 * of the clauses for chunk group filtering. For now, just assume the
+	 * selectivity of each clause is 0.50. In the future, we can use
+	 * correlation statistics or even just read the metadata directly (though
+	 * the join clauses will have Params and be harder to estimate).
+	 */
+	stripesToRead *= pow(0.50, nClauses);
```

Comment:
I think that makes sense. In future, maybe we can slightly increase this constant but I don't see an immediate problem with choosing `0.50`.

However, should we add a few simple test to `columnar_paths.sql` when `columnar.enable_qual_pushdown` is `ON` ?

Reply:
Improved this substantially.

Reply:
Improved costing a lot here. Also added tests in columnar_paths with pushdown=true.